File: 40-combination-sum-ii/40-combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        candidates.sort()
        
        def dfs(i, total, subset):
            if total==target:
                ans.append(subset[::])
                return
            if total>target or i>=len(candidates):
                return
            subset.append(candidates[i])
            dfs(i+1, total+candidates[i], subset)
            subset.pop()
            j=i
            while j<len(candidates) and candidates[j]== candidates[i]:
                j+=1
            dfs(j, total, subset)
            
        dfs(0, 0, [])
        return ans

File: 40-combination-sum-ii/40-combination-sum-ii.java
class Solution {
    List<List<Integer>> ans;
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        ans= new ArrayList<>();
        Arrays.sort(candidates);
        dfs(candidates, 0, target, new ArrayList<>());
        return ans;
    }
    void dfs(int[] candidates, int i, int target, List<Integer> list){
        if(target==0){
            ans.add(new ArrayList<>(list));
            return;
        }
        if(target<0 || i>=candidates.length) return;
        list.add(candidates[i]);
        dfs(candidates, i+1, target-candidates[i], list);
        list.remove(list.size()-1);
        int j=i;
        while(j<candidates.length && candidates[j]==candidates[i]) j++;
        dfs(candidates, j, target, list);
    }
}